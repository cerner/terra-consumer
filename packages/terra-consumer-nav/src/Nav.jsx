import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Button from 'terra-button';
import IconClose from 'terra-icon/lib/icon/IconClose';
import Popup from 'terra-popup';
import ResponsiveElement from 'terra-responsive-element';
import NavItems from './components/nav-items/NavItems';
import Modal from './components/modal/Modal';
import NavLogo from './components/nav-logo/NavLogo';
import NavHelp from './components/nav-help/NavHelp';
import NavBurgerButton from './components/nav-burger-button/NavBurgerButton';
import styles from './Nav.scss';
import UserProfile from './components/user-profile/UserProfile';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * An array of objects to be displayed as nav link options.
   */
  // eslint-disable-next-line react/forbid-prop-types
  navItems: PropTypes.array,
  /**
  * Object representing all the profile information
  */
  profile: PropTypes.shape({
    signinUrl: PropTypes.string,
    avatar: PropTypes.element,
  }),
  /**
   * An object defining the logo to be displayed
   */
  logo: PropTypes.shape({
    /**
     * The location where the image to be displayed is stored.
     */
    path: PropTypes.string,
    /**
     * Alternate text used be screen readers.
     */
    altText: PropTypes.string,
    /**
     * Whether or not the logo should be placed on top of a white card.
     */
    isCard: PropTypes.bool,
  }),
  /**
   * Callback function: should be used to close the nav on mobile devices.
   */
  onRequestClose: PropTypes.func.isRequired,
};

const defaultProps = {
  navItems: [],
  logo: {},
};

class Nav extends React.Component {
  constructor() {
    super();

    this.state = {
      isModalOpen: false,
      modalContent: {
        title: '',
        content: <div />,
      },
    };

    this.toggleModal = this.toggleModal.bind(this);
    this.handleOpenProfile = this.handleOpenProfile.bind(this);
    this.handleProfileLinkClick = this.handleProfileLinkClick.bind(this);
  }

  handleOpenProfile(modalContent, numberOfLinks) {
    this.toggleModal(modalContent, numberOfLinks);
  }

  handleProfileLinkClick() {
    // close our modal so they can see the new page
    this.setState({
      isModalOpen: false,
    });
    // Close the navigation panel as well
    this.props.onRequestClose();
  }

  toggleModal(modalObject) {
    if (modalObject.title && modalObject.content) {
      this.setState({
        modalContent: modalObject ||
          {
            title: '',
            content: <div />,
          },
      });
    }
    this.setState({
      isModalOpen: !this.state.isModalOpen,
    });
  }

  render() {
    const {
      navItems, profile, logo, onRequestClose, ...customProps
    } = this.props;
    const profileId = 'profile-popup-button';
    const defaultElement = (
      <Modal
        isModalOpen={this.state.isModalOpen}
        title={this.state.modalContent.title}
        onRequestClose={this.toggleModal}
      >
        {this.state.modalContent.content}
      </Modal>
    );

    const popup = (
      <Popup
        isOpen={this.state.isModalOpen}
        onRequestClose={this.toggleModal}
        targetRef={() => document.getElementById(profileId)}
        contentWidth="240"
        contentHeight="auto"
        contentAttachment="top right"
        isArrowDisplayed
      >
        {this.state.modalContent.content}
      </Popup>
    );

    return (
      <div
        {...customProps}
        id="terra-consumer-nav"
        className={cx('nav', { 'modal-open': this.state.isModalOpen }, customProps.className)}
      >
        <div className={cx('close-button-container')}>
          <Button icon={<IconClose />} className={cx('close-button')} onClick={() => { onRequestClose(); }} variant="utility" text="Close" />
        </div>
        <NavLogo {...logo} />
        <NavItems navItems={navItems} handleClick={onRequestClose} />
        <ResponsiveElement responsiveTo="window" defaultElement={defaultElement} medium={popup} />
        { profile &&
          <div className={cx('profile')}>
            <UserProfile
              {...profile}
              id={profileId}
              handleClick={this.handleOpenProfile}
              onLinkClick={this.handleProfileLinkClick}
              isSignIn={profile.signinUrl && !(profile.avatar || profile.userName)}
            />
          </div>
        }
      </div>
    );
  }
}

Nav.propTypes = propTypes;
Nav.defaultProps = defaultProps;
Nav.Help = NavHelp;
Nav.Burger = NavBurgerButton;
Nav.UserProfile = UserProfile;

export default Nav;
