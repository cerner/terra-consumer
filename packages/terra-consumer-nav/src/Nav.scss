@import url('https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700,900');
@import './variables';

:local {
  /* stylelint-disable selector-max-compound-selectors, max-nesting-depth */
  @media screen and (max-width: $terra-medium-breakpoint) {
    .mobile-panel {
      &::before {
        background: var(--terra-consumer-mobile-panel-background);
        background-attachment: fixed;
        background-size: cover;
        content: '';
        height: 100%;
        position: absolute;
        top: 0;
        width: var(--terra-consumer-mobile-panel-width, 320px);
        z-index: 100;

        @media screen and (max-width: $terra-tiny-breakpoint) {
          width: 100%;
        }
      }
    }
  }

  .container {
    min-height: 100%;
  }

  .nav {
    background: var(--terra-consumer-nav-background, rgba(255, 255, 255, 0.25));
    color: var(--terra-consumer-nav-text-color);
    /* nav is relative positioned and its height depends on the height of the main content next to it.
     * height: 100% makes sure nav height >= content height
    */
    height: 100%;
    // makes sure nav cover the window height on IE if the content next to it is shorter than window height
    min-height: 100vh;
    overflow-y: auto;
    padding-bottom: 60px;
    padding-top: var(--terra-consumer-nav-padding-top, 25px);
    position: relative;
    top: 0;
    user-select: none;
    width: var(--terra-consumer-nav-width, 320px);
    z-index: 2;
    @media screen and (max-width: $terra-tiny-breakpoint) {
      width: 100vw;
    }

    @media screen and (max-width: $terra-medium-breakpoint) {
      overflow: auto;
      padding-top: var(--terra-consumer--mobile-nav-padding-top, 0);
    }
  }

  .nav.no-profile {
    padding-bottom: 0;
  }

  @media screen and (max-width: $terra-medium-breakpoint) {
    .nav[aria-hidden='false'] {
      background: var(--terra-consumer-mobile-nav-background, rgba(255, 255, 255, 0.25));
      box-shadow: 0 1px 3px #383f42;
      display: block;
      height: 100%;
      z-index: 100;
    }

    .nav[aria-hidden='true'] {
      display: none;
    }
  }

  .close-button {
    display: none;
    @media screen and (max-width: $terra-medium-breakpoint) {
      background: var(--terra-consumer-mobile-close-button-background, none);
      border: 0;
      color: var(--terra-consumer-mobile-close-button-color, #fff);
      display: block;
      font-size: 16px;
      line-height: 16px;
      margin-right: 15px;
      margin-top: 15px;
      padding: 0;
      position: absolute;
      right: 0;
      top: 0;
    }

    &:hover,
    &:focus {
      background: var(--terra-consumer-mobile-close-button-hover-background, none);
      border-radius: 3px;
      color: var(--terra-consumer-mobile-close-button-hover-color, #ccc);
    }
  }
}
