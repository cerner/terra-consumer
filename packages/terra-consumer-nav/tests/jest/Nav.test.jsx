import React from 'react';
import Nav from '../../src/Nav';

const testData = {
  navItems: [{
    url: '#test',
    text: 'test',
    isActive: true,
  }],
  logo: {
    url: '',
    altText: 'test',
    isCard: false,
  },
  profile: {
    profileLinks: [{
      text: 'Account',
      url: 'http://localhost:8080/',
    },
    {
      text: 'Notifications',
      url: 'http://localhost:8080/',
    }],
    userName: 'John Snow',
    profileId: 'profile-popup-button',
    signoutUrl: 'http://localhost:8080/',
  },
  isMobileNavOpen: false,
  onRequestClose: () => {},
  currentURL: 'http://localhost:8080/'
};

describe('Nav', () => {
  const defaultRender = <Nav {...testData} />;
  const wrapper = shallow(defaultRender);

  // Snapshot Tests
  it('should render a default component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // Structure Tests
  it('should have the id terra-consumer-nav', () => {
    expect(wrapper.prop('id')).toContain('terra-consumer-nav');
  });

  it('should have have the profile component', () => {
    expect(wrapper.find('.profile').exists()).toEqual(true);
  });
  
  it('should have the currentURL prop', () => {
    expect(wrapper.find('NavItems').first().props().currentURL).toEqual('http://localhost:8080/');
  });
});

describe('Nav without profile', () => {
  const { profile, currentURL, ...data } = testData;
  const wrapper = shallow(<Nav {...data} />);

  // Snapshot Tests
  it('should render a nav without a profile component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  // Structure Tests
  it('should have not have the profile component', () => {
    expect(wrapper.find('.profile').exists()).toEqual(false);
  });
  
  it('should not have the currentURL prop', () => {
    expect(wrapper.find('NavItems').first().props().currentURL).toEqual('http://localhost:8080/');
  });
});
