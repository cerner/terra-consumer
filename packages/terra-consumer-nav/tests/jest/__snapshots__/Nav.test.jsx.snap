// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render a default component 1`] = `
<div
  className="nav"
  id="terra-consumer-nav"
  isMobileNavOpen={false}
>
  <div
    className="close-button-container"
  >
    <Button
      className="close-button"
      icon={
        <IconClose
          data-name="Layer 1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />
      }
      isBlock={false}
      isCompact={false}
      isDisabled={false}
      isReversed={false}
      onClick={[Function]}
      type="button"
      variant="link"
    />
  </div>
  <NavLogo
    altText="test"
    isCard={false}
    url=""
  />
  <NavItems
    currentURL="http://localhost:8080/"
    handleClick={[Function]}
    navItems={
      Array [
        Object {
          "isActive": true,
          "text": "test",
          "url": "#test",
        },
      ]
    }
  />
  <ResponsiveElement
    defaultElement={
      <Modal
        isModalOpen={false}
        onRequestClose={[Function]}
        title=""
      >
        <div />
      </Modal>
    }
    medium={
      <Popup
        attachmentBehavior="auto"
        boundingRef={null}
        classNameArrow={null}
        classNameContent={null}
        classNameOverlay={null}
        contentAttachment="top right"
        contentHeight="auto"
        contentWidth="240"
        isArrowDisplayed={true}
        isContentFocusDisabled={false}
        isHeaderDisabled={false}
        isOpen={false}
        onRequestClose={[Function]}
        targetRef={[Function]}
      >
        <div />
      </Popup>
    }
    responsiveTo="window"
  />
  <div
    className="profile"
  >
    <InjectIntl(UserProfile)
      handleClick={[Function]}
      id="profile-popup-button"
      onLinkClick={[Function]}
      profileId="profile-popup-button"
      profileLinks={
        Array [
          Object {
            "text": "Account",
            "url": "http://localhost:8080/",
          },
          Object {
            "text": "Notifications",
            "url": "http://localhost:8080/",
          },
        ]
      }
      signoutUrl="http://localhost:8080/"
      userName="John Snow"
    />
  </div>
</div>
`;

exports[`Nav without profile should render a nav without a profile component 1`] = `
<div
  className="nav"
  id="terra-consumer-nav"
  isMobileNavOpen={false}
>
  <div
    className="close-button-container"
  >
    <Button
      className="close-button"
      icon={
        <IconClose
          data-name="Layer 1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        />
      }
      isBlock={false}
      isCompact={false}
      isDisabled={false}
      isReversed={false}
      onClick={[Function]}
      type="button"
      variant="link"
    />
  </div>
  <NavLogo
    altText="test"
    isCard={false}
    url=""
  />
  <NavItems
    handleClick={[Function]}
    navItems={
      Array [
        Object {
          "isActive": true,
          "text": "test",
          "url": "#test",
        },
      ]
    }
  />
  <ResponsiveElement
    defaultElement={
      <Modal
        isModalOpen={false}
        onRequestClose={[Function]}
        title=""
      >
        <div />
      </Modal>
    }
    medium={
      <Popup
        attachmentBehavior="auto"
        boundingRef={null}
        classNameArrow={null}
        classNameContent={null}
        classNameOverlay={null}
        contentAttachment="top right"
        contentHeight="auto"
        contentWidth="240"
        isArrowDisplayed={true}
        isContentFocusDisabled={false}
        isHeaderDisabled={false}
        isOpen={false}
        onRequestClose={[Function]}
        targetRef={[Function]}
      >
        <div />
      </Popup>
    }
    responsiveTo="window"
  />
</div>
`;
